File: src/main/java/org/parsimonygroup/FunctionFilterBootstrap.java
package org.parsimonygroup;

import cascading.flow.FlowProcess;
import cascading.operation.BaseOperation;
import cascading.operation.Function;
import cascading.operation.FunctionCall;
import cascading.tuple.Fields;
import cascading.tuple.Tuple;
import cascading.tuple.TupleEntry;
import cascading.tuple.TupleEntryCollector;
import clojure.lang.IFn;

public class FunctionFilterBootstrap extends BaseOperation implements Function {
  private IFn reader;
  private IFn function;
  private ClojureCascadingHelper clojureHelper;
  private IFn writer;
  private IFn cljCallback;

  public FunctionFilterBootstrap(Fields inFields, Fields outFields, IFn reader,IFn writer, IFn function, IFn cljCallback, String fnNsName) {
    super(inFields.size(),outFields);
    this.reader = reader;
    this.function = function;
    this.writer = writer;
    this.cljCallback = cljCallback;
    this.clojureHelper = new ClojureCascadingHelper(fnNsName);
  }

  public void operate(FlowProcess flowProcess, FunctionCall functionCall) {
    try {
      clojureHelper.bootClojure();
    } catch (Exception e) {
      e.printStackTrace();
    }
    TupleEntry arguments = functionCall.getArguments();
    Tuple result = processData(arguments);

    TupleEntryCollector outputCollector = functionCall.getOutputCollector();
    outputCollector.add(result);
  }

  private Tuple processData(TupleEntry arguments) {
    Tuple result = new Tuple();
    try {
      if((Boolean)clojureHelper.filterCall(arguments, function, cljCallback, reader, writer)) {
        result.add(arguments.getTuple().get(arguments.getFields().size() -1));
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
    return result;
  }
}

File: src/main/clojure/org/parsimonygroup/function_filter_bootstrap.clj
(ns org.parsimonygroup.function-filter-bootstrap
  (:import [org.parsimonygroup ClojureCascadingHelper]
           [cascading.flow FlowProcess]
           [cascading.operation BaseOperation Function FunctionCall]
           [cascading.tuple Fields Tuple TupleEntryCollector TupleEntry])
  (:use org.parsimonygroup.utils))

(gen-class
 :name org.parsimonygroup.FunctionFilterBootstrapInClojure
 :extends cascading.operation.BaseOperation
 :implements [cascading.operation.Function]
 :constructors {[cascading.tuple.Fields cascading.tuple.Fields
                 clojure.lang.IFn clojure.lang.IFn
                 clojure.lang.IFn clojure.lang.IFn String]
                [int cascading.tuple.Fields]}
 :init init
 :state state)


;; Oddly Cascading gags when trying to serialize
;; Clojure keywords so we can't use them here:
(defn -init [in-fields out-fields reader writer function
             clj-callback fn-ns-name]
  [[(.size in-fields) out-fields]
   {"reader" reader "writer" writer "function" function
    "clj-callback" clj-callback "fn-ns-name" fn-ns-name}])

(defn -prepare [this _ _]
  (boot-clojure (.state this)))

(defn process-data [this arguments]
   (let [state (.state this)]
     (if ((state "clj-callback" )
          (state "reader" )
          (state "function" )  (iterator-seq (.. arguments getTuple iterator)))
      (.. arguments getTuple))))


(defn -operate [this flow-process function-call]
  (let [arguments (.getArguments function-call)
        result (process-data this arguments)
        output-collector (.getOutputCollector function-call)]
    (if (not (nil? result)) (.add output-collector result))))

